go/database/mpt/io/interrupt_test.go

```diff
@@ -0,0 +1,168 @@
+			}
+
+			// save max count and reset number of writes
+			maxCount := writer.numOfWrites.Load()
+			// reset number of writes, so we can compare
+			// that export was indeed interrupted
+			writer.numOfWrites = atomic.Uint32{}
+			writer.signalWrite = true
+			// give the interrupt some time
+			writer.waitTime = 100 * time.Millisecond
+			hasEnded := sync.WaitGroup{}
+			hasEnded.Add(1)
+			go func() {
+
+				defer hasEnded.Done()
+				if err := tf.export(sourceDir, writer); err != nil {
+					got := err.Error()
+					want := errCanceled.Error()
+					if !strings.Contains(got, want) {
+						t.Errorf("unexpected error: got: %v, want: %v", got, want)
+					}
+					return
+				}
+				t.Error("export was interrupt hence should raise an error")
+			}()
+
+			// wait for first write
+			writerWg.Wait()
+			err := syscall.Kill(syscall.Getpid(), syscall.SIGINT)
+			if err != nil {
+				t.Fatal("failed to create a SIGINT signal")
+			}
+
+			// wait for the export to finish
+			hasEnded.Wait()
+			if maxCount == writer.numOfWrites.Load() {
+				t.Error("export was not interrupted")
+			}
+
+			// lastly check that the database is not corrupted
+			tf.check(t, sourceDir)
+		})
+	}
+}
+
+func createTestLive(t *testing.T, sourceDir string) {
+	t.Helper()
+	db := createExampleLiveDB(t, sourceDir)
+	if err := db.Close(); err != nil {
+		t.Fatalf("failed to close example live db: %v", err)
+	}
+}
+
+func createTestArchive(t *testing.T, sourceDir string) {
+	t.Helper()
+	source, err := mpt.OpenArchiveTrie(sourceDir, mpt.S5ArchiveConfig, 1024)
+	if err != nil {
+		t.Fatalf("failed to create archive: %v", err)
+	}
+	fillTestBlocksIntoArchive(t, source)
+	if err = source.Close(); err != nil {
+		t.Fatalf("failed to close test DB: %v", err)
+	}
+}
+
+// checkCanOpenLiveDB makes sure LiveDB is not corrupted and can be opened (and closed)
+func checkCanOpenLiveDB(t *testing.T, sourceDir string) {
+	db, err := mpt.OpenGoFileState(sourceDir, mpt.S5LiveConfig, mpt.DefaultMptStateCapacity)
+	if err != nil {
+		t.Fatalf("failed to open db: %v", err)
+	}
+	err = db.Close()
+	if err != nil {
+		t.Fatalf("failed to close db: %v", err)
+	}
+}
+
+// checkCanOpenLiveDB makes sure Archive is not corrupted and can be opened (and closed)
+func checkCanOpenArchive(t *testing.T, sourceDir string) {
+	archive, err := mpt.OpenArchiveTrie(sourceDir, mpt.S5ArchiveConfig, mpt.DefaultMptStateCapacity)
+	if err != nil {
+		t.Fatalf("failed to open archive: %v", err)
+	}
+	err = archive.Close()
+	if err != nil {
+		t.Fatalf("failed to close archive: %v", err)
+	}
+}
+
+type mockWriter struct {
+	numOfWrites atomic.Uint32
+	signalWrite bool
+	waitTime    time.Duration
+	wg          *sync.WaitGroup
+}
+
+func (m *mockWriter) Write([]byte) (n int, err error) {
+	m.numOfWrites.Add(1)
+	// inform the test that first write has happened
+	if m.signalWrite {
+		m.signalWrite = false
+		m.wg.Done()
+	}
+	// slow down writing to be able to interrupt the export
+	time.Sleep(m.waitTime)
```

Comment:
The use of timing may make the test flaky and inefficient. 

How about triggering the interrupt signal from within the writer after a given number of writes? This would make the test resilient against relative timing. Also, you can simplify the test since you no longer would need multiple go-routines for testing.

Reply:
I implemented it but it still requires some `sleep` time. The export is too fast... even tho `cancel()` is called the func call ends before the interrupt is noticed. Another option would be to create a bigger database to extend the export.